File: cosmic_fusion.py
# services/cosmic_fusion.py

from typing import Dict, Any
from services.astrology_service import AstrologyService
from services.numerology_service import NumerologyService
#from services.feedback_service import FeedbackService
from datetime import datetime
import math

class CosmicFusionService:

    @staticmethod
    def normalize(value: float) -> float:
        return max(0, min(10, value))

    @staticmethod
    def generate_cosmic_profile(sun_sign: str, destiny: int, soul_urge: int, personality: int, life_path: int, birthday: int) -> Dict:
        key_traits = {
            "destiny": destiny,
            "soulUrge": soul_urge,
            "personality": personality
        }

        archetype_summary = AstrologyService.get_archetype_summary(sun_sign)

        love_energy = destiny * 0.4 + personality * 0.4 + AstrologyService.get_love_influence(sun_sign) * 0.2
        career_energy = destiny * 0.5 + soul_urge * 0.3 + AstrologyService.get_career_influence(sun_sign) * 0.2
        health_energy = personality * 0.5 + AstrologyService.get_health_influence(sun_sign) * 0.5
        finance_energy = destiny * 0.6 + soul_urge * 0.2 + AstrologyService.get_finance_influence(sun_sign) * 0.2
        spiritual_energy = soul_urge * 0.5 + AstrologyService.get_spiritual_influence(sun_sign) * 0.5

        energy_vectors = {
            "love": CosmicFusionService.normalize(love_energy),
            "career": CosmicFusionService.normalize(career_energy),
            "health": CosmicFusionService.normalize(health_energy),
            "finance": CosmicFusionService.normalize(finance_energy),
            "spiritual": CosmicFusionService.normalize(spiritual_energy),
        }

        energy_vectors["overall"] = CosmicFusionService.normalize(
            sum(energy_vectors.values()) / 5
        )

        birthday_factor = birthday % 10
        cosmic_alignment = CosmicFusionService.normalize(
            (energy_vectors["overall"] + birthday_factor) / 2
        )

        return {
            "fusedInsights": {
                "cosmicAlignment": cosmic_alignment,
                "archetypeSummary": archetype_summary,
                "keyTraits": key_traits,
                "energyVectors": energy_vectors
            },
            "sunSign": sun_sign,
            "numerologyCore": key_traits,
            "birthday": birthday
        }

    @staticmethod
    def generate_enhanced_daily_insight(cosmic_profile: Dict, target_date: str, learning_insights: Dict[str, Any]) -> Dict:
        base = cosmic_profile["fusedInsights"]["energyVectors"]
        sun_sign = cosmic_profile["sunSign"]
        destiny = cosmic_profile["numerologyCore"]["destiny"]

        date = datetime.strptime(target_date, "%Y-%m-%d")
        transits = CosmicFusionService.get_transit_modifiers(sun_sign, date)
        numerology_mods = CosmicFusionService.get_numerology_cycle_modifiers(destiny, date)

        confidence_weight = min(1, learning_insights.get("accuracy_metrics", {}).get("overall", 7) / 10)

        def combined(base_val, transit_mod, num_mod):
            return CosmicFusionService.normalize(
                base_val * (1 - confidence_weight) + ((transit_mod + num_mod) / 2) * confidence_weight
            )

        energy_readings = {
            "overall": combined(base["overall"], transits["overall"], numerology_mods["overall"]),
            "love": combined(base["love"], transits["love"], numerology_mods["love"]),
            "career": combined(base["career"], transits["career"], numerology_mods["career"]),
            "health": combined(base["health"], transits["health"], numerology_mods["health"]),
            "finance": combined(base["finance"], transits["finance"], numerology_mods["finance"]),
        }

        manifestation_power = round((energy_readings["overall"] * 0.6 + base["spiritual"] * 0.4) * confidence_weight)
        spiritual_focus = CosmicFusionService.get_spiritual_focus_message(base["spiritual"])
        optimal_activities = CosmicFusionService.get_optimal_activities(energy_readings)
        caution_areas = CosmicFusionService.get_caution_areas(energy_readings)

        return {
            "energyReadings": energy_readings,
            "manifestationPower": manifestation_power,
            "spiritualFocus": spiritual_focus,
            "optimalActivities": optimal_activities,
            "cautionAreas": caution_areas,
            "cosmicInfluences": [
                f"Your sun sign {sun_sign} supports new beginnings today.",
                f"Numerology destiny number {destiny} encourages persistence and focus."
            ],
            "personalizedGuidance": [
                "Stay open to unexpected opportunities.",
                "Practice mindfulness to enhance clarity."
            ]
        }

    @staticmethod
    def get_transit_modifiers(sun_sign: str, date: datetime) -> Dict[str, float]:
        day = date.day
        return {
            "overall": 5 + math.sin(day) * 2,
            "love": 5 + math.cos(day) * 2,
            "career": 5 + math.sin(day / 2) * 2,
            "health": 5 + math.cos(day / 3) * 2,
            "finance": 5 + math.sin(day / 4) * 2,
        }

    @staticmethod
    def get_numerology_cycle_modifiers(destiny: int, date: datetime) -> Dict[str, float]:
        day = date.day
        mod = 1 if day % 2 == 0 else -1
        base = 5
        return {
            "overall": base + mod,
            "love": base + mod * 0.8,
            "career": base + mod * 0.9,
            "health": base + mod * 0.7,
            "finance": base + mod * 0.85,
        }

    @staticmethod
    def get_spiritual_focus_message(spiritual_energy: float) -> str:
        if spiritual_energy >= 8:
            return "High spiritual awareness; ideal for meditation and inner work."
        elif spiritual_energy >= 5:
            return "Balanced spiritual energy; practice gratitude."
        elif spiritual_energy >= 3:
            return "Low spiritual energy; focus on grounding activities."
        else:
            return "Minimal spiritual energy today; prioritize rest and recuperation."

    @staticmethod
    def get_optimal_activities(energies: Dict[str, float]) -> list:
        activities = []
        if energies["overall"] >= 7:
            activities.append("Initiate new projects")
        if energies["love"] >= 7:
            activities.extend(["Romantic engagements", "Social bonding"])
        if energies["career"] >= 7:
            activities.extend(["Career advancement tasks", "Networking"])
        if energies["health"] >= 7:
            activities.extend(["Physical exercise", "Wellness activities"])
        if energies["finance"] >= 7:
            activities.extend(["Financial planning", "Investments"])
        return activities or ["Rest and reflection"]

    @staticmethod
    def get_caution_areas(energies: Dict[str, float]) -> list:
        cautions = []
        if energies["overall"] <= 3:
            cautions.extend(["Avoid big decisions", "Rest more"])
        if energies["love"] <= 3:
            cautions.append("Avoid relationship conflicts")
        if energies["career"] <= 3:
            cautions.append("Delay negotiations")
        if energies["health"] <= 3:
            cautions.append("Prioritize recovery")
        if energies["finance"] <= 3:
            cautions.append("Avoid risky financial moves")
        return cautions or ["Maintain balance and focus"]

File: services/cosmic-fusion.ts
// frontend/src/services/cosmic-fusion.ts

import { AstrologyService } from './astrology.js';
import { NumerologyService } from './numerology.js';
import { FeedbackService } from './feedback.js';

export interface CosmicProfile {
  fusedInsights: {
    cosmicAlignment: number; // 0-10 scale
    archetypeSummary: string;
    keyTraits: {
      destiny: number;
      soulUrge: number;
      personality: number;
    };
    energyVectors: {
      love: number;
      career: number;
      health: number;
      finance: number;
      spiritual: number;
      overall: number;
    };
  };
  sunSign: string;
  numerologyCore: {
    destiny: number;
    soulUrge: number;
    personality: number;
  };
  birthday: number;
}

export interface EnhancedDailyInsight {
  energyReadings: {
    overall: number;
    love: number;
    career: number;
    health: number;
    finance: number;
  };
  manifestationPower: number; // 0-10 scale
  spiritualFocus: string;
  optimalActivities: string[];
  cautionAreas: string[];
  cosmicInfluences: string[];
  personalizedGuidance: string[];
}

export class CosmicFusionService {

  // Fuse astrology + numerology + birthday into cosmic profile object
  static generateCosmicProfile(
    sunSign: string,
    destiny: number,
    soulUrge: number,
    personality: number,
    lifePath: number,
    birthday: number
  ): CosmicProfile {

    // 1. Calculate base numerology traits weighted average
    const keyTraits = {
      destiny,
      soulUrge,
      personality,
    };

    // 2. Astrological archetype influence (simplified: can extend with ephemeris data)
    const archetypeSummary = AstrologyService.getArchetypeSummary(sunSign);

    // 3. Energy vectors by combining numerology and astrology
    // Sample weights for traits (scale 1-10)
    const loveEnergy = (destiny * 0.4 + personality * 0.4 + AstrologyService.getLoveInfluence(sunSign) * 0.2);
    const careerEnergy = (destiny * 0.5 + soulUrge * 0.3 + AstrologyService.getCareerInfluence(sunSign) * 0.2);
    const healthEnergy = (personality * 0.5 + AstrologyService.getHealthInfluence(sunSign) * 0.5);
    const financeEnergy = (destiny * 0.6 + soulUrge * 0.2 + AstrologyService.getFinanceInfluence(sunSign) * 0.2);
    const spiritualEnergy = (soulUrge * 0.5 + AstrologyService.getSpiritualInfluence(sunSign) * 0.5);

    // Normalize to 0-10 scale
    const normalize = (val: number) => Math.min(10, Math.max(0, val));

    const energyVectors = {
      love: normalize(loveEnergy),
      career: normalize(careerEnergy),
      health: normalize(healthEnergy),
      finance: normalize(financeEnergy),
      spiritual: normalize(spiritualEnergy),
      overall: 0 // to compute next
    };
    energyVectors.overall = normalize(
      (energyVectors.love + energyVectors.career + energyVectors.health + energyVectors.finance + energyVectors.spiritual) / 5
    );

    // 4. Compute cosmic alignment score by blending energy overall and birthday numerology (e.g., birth day mod 10)
    // Example: birthday mod 10 scaled to 0-10, averaged with overall energy
    const birthdayFactor = (birthday % 10);
    const cosmicAlignment = normalize((energyVectors.overall + birthdayFactor) / 2);

    return {
      fusedInsights: {
        cosmicAlignment,
        archetypeSummary,
        keyTraits,
        energyVectors,
      },
      sunSign,
      numerologyCore: { destiny, soulUrge, personality },
      birthday,
    };
  }

  // Generate enhanced daily insight using cosmic profile + date + AI feedback
  static generateEnhancedDailyInsight(
    cosmicProfile: CosmicProfile,
    targetDate: string,
    learningInsights: any // object returned from FeedbackService with accuracy, sample counts etc
  ): EnhancedDailyInsight {

    // 1. Extract base energy vectors
    const baseEnergies = cosmicProfile.fusedInsights.energyVectors;

    // 2. Calculate modifiers from astrological transits (mock example)
    // Real implementation should query AstrologyService with targetDate for transits/aspects
    const transitModifiers = this.getTransitModifiers(cosmicProfile.sunSign, targetDate);

    // 3. Calculate numerology cycle modifiers for targetDate (e.g., personal year cycle)
    const numerologyModifiers = this.getNumerologyCycleModifiers(cosmicProfile.numerologyCore.destiny, targetDate);

    // 4. Combine base energy with modifiers weighted by AI feedback confidence
    const confidenceWeight = learningInsights?.accuracy_metrics?.overall
      ? Math.min(1, learningInsights.accuracy_metrics.overall / 10)
      : 0.7; // default confidence

    const combinedEnergy = (base: number, transitMod: number, numMod: number) => {
      return Math.min(
        10,
        Math.max(
          0,
          base * (1 - confidenceWeight) +
          ((transitMod + numMod) / 2) * confidenceWeight
        )
      );
    };

    const overall = combinedEnergy(baseEnergies.overall, transitModifiers.overall, numerologyModifiers.overall);
    const love = combinedEnergy(baseEnergies.love, transitModifiers.love, numerologyModifiers.love);
    const career = combinedEnergy(baseEnergies.career, transitModifiers.career, numerologyModifiers.career);
    const health = combinedEnergy(baseEnergies.health, transitModifiers.health, numerologyModifiers.health);
    const finance = combinedEnergy(baseEnergies.finance, transitModifiers.finance, numerologyModifiers.finance);

    // 5. Manifestation power derived from overall and spiritual energies, weighted by feedback
    const manifestationPower = Math.round(
      (overall * 0.6 + baseEnergies.spiritual * 0.4) * confidenceWeight
    );

    // 6. Spiritual focus - generate dynamic message based on spiritual energy
    const spiritualFocus = this.getSpiritualFocusMessage(baseEnergies.spiritual);

    // 7. Recommended activities and caution areas based on thresholds
    const optimalActivities = this.getOptimalActivities({ overall, love, career, health, finance });
    const cautionAreas = this.getCautionAreas({ overall, love, career, health, finance });

    // 8. Cosmic influences and personalized guidance placeholders (expand with NLP/AI in future)
    const cosmicInfluences = [
      `Your sun sign ${cosmicProfile.sunSign} supports new beginnings today.`,
      `Numerology destiny number ${cosmicProfile.numerologyCore.destiny} encourages persistence and focus.`
    ];

    const personalizedGuidance = [
      'Stay open to unexpected opportunities.',
      'Practice mindfulness to enhance clarity.'
    ];

    return {
      energyReadings: { overall, love, career, health, finance },
      manifestationPower,
      spiritualFocus,
      optimalActivities,
      cautionAreas,
      cosmicInfluences,
      personalizedGuidance,
    };
  }

  // Helpers:

  private static getTransitModifiers(sunSign: string, targetDate: string) {
    // Placeholder for actual astrology transit computations
    // Return modifiers in 0-10 scale for energy categories
    // Simulate mild daily variation
    return {
      overall: 5 + Math.sin(new Date(targetDate).getDate()) * 2,
      love: 5 + Math.cos(new Date(targetDate).getDate()) * 2,
      career: 5 + Math.sin(new Date(targetDate).getDate() / 2) * 2,
      health: 5 + Math.cos(new Date(targetDate).getDate() / 3) * 2,
      finance: 5 + Math.sin(new Date(targetDate).getDate() / 4) * 2,
    };
  }

  private static getNumerologyCycleModifiers(destinyNumber: number, targetDate: string) {
    // Placeholder: calculate numerology personal year or pinnacle cycles
    // For now, cycle modulates energies by +/- 1 based on day parity
    const day = new Date(targetDate).getDate();
    const mod = (day % 2 === 0) ? 1 : -1;
    const base = 5;

    return {
      overall: base + mod,
      love: base + mod * 0.8,
      career: base + mod * 0.9,
      health: base + mod * 0.7,
      finance: base + mod * 0.85,
    };
  }

  private static getSpiritualFocusMessage(spiritualEnergy: number): string {
    if (spiritualEnergy >= 8) return 'High spiritual awareness; ideal for meditation and inner work.';
    if (spiritualEnergy >= 5) return 'Balanced spiritual energy; practice gratitude.';
    if (spiritualEnergy >= 3) return 'Low spiritual energy; focus on grounding activities.';
    return 'Minimal spiritual energy today; prioritize rest and recuperation.';
  }

  private static getOptimalActivities(energies: Record<string, number>): string[] {
    const activities: string[] = [];
    if (energies.overall >= 7) activities.push('Initiate new projects');
    if (energies.love >= 7) activities.push('Romantic engagements', 'Social bonding');
    if (energies.career >= 7) activities.push('Career advancement tasks', 'Networking');
    if (energies.health >= 7) activities.push('Physical exercise', 'Wellness activities');
    if (energies.finance >= 7) activities.push('Financial planning', 'Investments');
    if (activities.length === 0) activities.push('Rest and reflection');
    return activities;
  }

  private static getCautionAreas(energies: Record<string, number>): string[] {
    const cautions: string[] = [];
    if (energies.overall <= 3) cautions.push('Avoid big decisions', 'Rest more');
    if (energies.love <= 3) cautions.push('Avoid relationship conflicts');
    if (energies.career <= 3) cautions.push('Delay negotiations');
    if (energies.health <= 3) cautions.push('Prioritize recovery');
    if (energies.finance <= 3) cautions.push('Avoid risky financial moves');
    if (cautions.length === 0) cautions.push('Maintain balance and focus');
    return cautions;
  }
}
